distutils/file_util.py

```diff
@@ -70,9 +70,8 @@ def copy_file(  # noqa: C901
     verbose=1,
     dry_run=0,
 ):
-    """Copy a file 'src' to 'dst'.  If 'dst' is a directory, then 'src' is
-    copied there with the same name; otherwise, it must be a filename.  (If
-    the file exists, it will be ruthlessly clobbered.)  If 'preserve_mode'
+    """Copy a file 'src' to 'dst'.
+    (If the file exists, it will be ruthlessly clobbered.)  If 'preserve_mode'
```

Comment:
I don't think we can assume that distutils is the only consumer of this function. Maybe no one else is using it, but we can't simply remove it without risking breaking users. Let's do this instead:

1. Rename `copy_file` to `_copy_file_direct`, implementing this new, limited interface.
2. Create a new `copy_file` that implements the backward-compatible behavior (infer dst from dir),  raises a DeprecationWarning, and then calls `_copy_file_direct`.
3. Retain the logic of emitting only the dirname if the basenames match (I don't think that change is necessary to get the performance gain).
4. Create a new `_copy_file_to_dir` that assumes the dst is a dir and performs the join and basename, calling `_copy_file_direct`.

What do you think?

Reply:
for 3. the logic is broken as far as I can tell. we can't keep the syscall to `isdir()` to determine whether we're dealing with a directory because that's what makes the function slow. 
for context, this function is called for thousands of files, it's used by distutils and setuptools/pip when working with packages. (the profiling run was for `python setup.py bdist_wheel`)

I think we can't have a replacement function be private with a underscore. It's a public API that is used in other places. it think linters will complain if we import a private _ function in other places (note that setuptools is vendoring this repo and might have more CI rules).

searching on github `distutils.file_util AND copy_file AND path:.py`, it's finding 2k files using this function. Going over the results, I see a few are passing a directory. That is problematic. :(
I don't think adding a warning would be helpful, distutils is removed in python 3.12 and already raising warnings/errors. If people ain't fixing errors because the module is removed, they're not going to care about yet another warning. ^^

what I can think of?
* make two functions copy_file_direct() and copy_file_to_dir(). 
* fix code inside of distutils to use the appropriate function.
* keep copy_file() logic to check whether the argument is a directory + call one of the 2 functions.

Reply:
I've started the refactoring... but unfortunately found out `copy_file()` is wrapped inside of `distutils.cmd::Command` class and there are a dozen of invocations through `self.copy_file` across the codebase. 
**If we don't want to fix copy_file()** but would rather ship a copy_file_direct() + copy_file_to_dir(), then we would have to expose both functions though the class and adjust a dozen invocations. it's not hard to do but that's a fair amount of boilerplate and code to refactor. 
I would like to have confirmation we want to do that before engaging into that.

realistically, `distutils` is removed from the interpreter as of python 3.12, it's only available inside of setuptools where it's vendored. I initially made the PR on setuptools where it works and pass all tests, it provides a welcome performance improvement to package operations.

Is there really a risk in fixing copy_file() to break other packages? packages that are not already broken because distutils is removed.

Reply:
Hyrum's law usually hints hard on distutils/setuptools...

If you want to define 2 other new functions (one for files or one for directories) it should be fine. But my opinion is that we should keep the old behaviour (maybe behind a deprecated warning) for a couple of years before dropping it.

Reply:
I agree, that sounds messy. Let's consider some other options.